### ISSUE TITLE
Pathless <Route> inside of <Route children> breaks when parent route's path does not match

### ISSUE THREAD

**Issue by firasdib** [ReactTraining/react-router#4695]
Not entirely sure where to post this so giving it a try here.

Is there a reason `withRouter` uses `render` instead of `children`? Would it not make more sense to use `children`, since its rendering a route under the hood we should always render it and not impose any limitations on it?

**Comment by firasdib** [comment]
Now that I think about it, this would break if the component was not rendered by another route.

My issue is that I have a component that is always rendered using `children` which animates in/out. The component being rendered uses `withRouter` which unmounts immediately for some reason, instead of remaining mouted, preventing me from doing the leave animation.

**Comment by pshrmn** [comment]
It renders a pathless `<Route>`, which means it always matches (so `render` is the same as `children`).

**Comment by ryanflorence** [comment]
Closing because I'm confident pathless routes don't cause remounting, we'd need a test case proving otherwise. Something else is going on in your app to cause that. If you're sure its our problem, please make a test case whether on codepen or as a failing test in the repo. Thanks :)

**Comment by firasdib** [comment]
I was using it like adviced, `export default withRouter(component)` . Removing the `withRouter` part fixed it for me so I'll live with those implications for now.

**Comment by pshrmn** [comment]
@ryanflorence I believe that this is actually a bug emanating from f90d753.

**Comment by felixmosh** [comment]
Have the same issue, 
```jsx
<Route path='/' children={() => (<SomeComponentWrappedWithRouter />)}/>
```
When navigating to `/xxx/`, SomeComponentWrappedWithRouter won't mount.

**Comment by Shrugs** [comment]
Not perfect, but one workaround is to set `computedMatch` to something truthy so render is always called. Just, like, don't actually use that value within `render` 😂 

```
import React from 'react'
import { Route } from 'react-router-dom'

const alwaysRenderWithRouter = (Component) => {
  const C = (props) => {
    const { ...remainingProps } = props
    return (
      <Route computedMatch={{}} render={routeComponentProps => (
        <Component {...remainingProps} {...routeComponentProps} />
      )} />
    )
  }

  C.displayName = `alwaysRenderWithRoute(${Component.displayName || Component.name})`
  C.WrappedComponent = Component

  return C
}

export default alwaysRenderWithRouter

```

**Comment by piotrblasiak** [comment]
Is the recommended workaround still what @Shrugs suggested, and/or is there any other fix?

**Comment by penx** [comment]
I wrote a fix for #5141 (#5142) which was closed as a duplicate of this ticket. The PR has been open for 9 months without feedback, other than "they are not ready yet" loosely relating to this in https://github.com/ReactTraining/react-router/issues/5852

Any chance on some feedback on this PR?

**Comment by DylanVann** [comment]
I spent so long on this bug 😞. Everything I read kept saying I had an update blocking issue and I should use `withRouter`.
